### PR TITLE
Support for OData $count operations as per v4.01 spec section 11.2.10

### PR DESCRIPTION
**Pull request by joshcomley**
### Issues

I am unable to find an open issue relating to this.

### Description

As per the title, this is to implement support for $count operations such as:

    GET http://host/service/Products/$count
    GET http://host/service/Products/$count?$filter=Price lt 10.00
    GET http://host/service/Customers?$filter=Interests/$count gt 5
    GET http://host/service/Categories?$filter=Products/$count($filter=Price gt 5.0) gt 1

With this change, the following is still unsupported:

    GET http://host/service/Products/$filter=@foo/$count?@foo=Price lt 10.00

Specification reference:
http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_RequestingtheNumberofItemsinaCollect

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added* (although only URI builder tests)
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

Functional tests need to be written.

Unsure if docs need to be updated.

### CODE REVIEW

**Comment by xuzhg** [comment]
@joshcomley Would you please file two issues:
1. One issue is to describe the functionality/requirement/problem/etc implemented in this PR.
2. The other is for the remaining unsupported scenario

**Comment by xuzhg** [comment]
@joshcomley I see there are some build errors:

```
src\Microsoft.OData.Core\JsonLight\ODataJsonLightWriter.cs(562,69): Error CS1026: ) expected
src\Microsoft.OData.Core\JsonLight\ODataJsonLightWriter.cs(562,84): Error CS1002: ; expected
src\Microsoft.OData.Core\JsonLight\ODataJsonLightWriter.cs(562,84): Error CS1513: } expected
src\Microsoft.OData.Core\JsonLight\ODataJsonLightWriter.cs(562,86): Error CS1525: Invalid expression term '||'
src\Microsoft.OData.Core\JsonLight\ODataJsonLightWriter.cs(562,115): Error CS1002: ; expected
src\Microsoft.OData.Core\JsonLight\ODataJsonLightWriter.cs(562,115): Error CS1513: } expected
Process 'msbuild.exe' exited with code '1'.
```
Would you please help take a look?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriBuilder/CountBuilderTests.cs [hunk 1]
```diff
@@ -11,6 +11,55 @@ namespace Microsoft.OData.Tests.ScenarioTests.UriBuilder
...
+
+        [Fact]
+        public void ExpandCountOnChildCollectionWorks()
+        {
+            Uri queryUri = new Uri("People?$expand=MyPaintings/$count", UriKind.Relative);
+            Uri actualUri = UriBuilder(queryUri, ODataUrlKeyDelimiter.Parentheses, settings);
+            Assert.Equal(new Uri("http://gobbledygook/People?$expand=MyPaintings%2F%24count"), actualUri);
+        }
+
+        // This is not supported, yet
```
Comment (biaol-odata) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ScenarioTests/UriBuilder/CountBuilderTests.cs]: We need to support parameter alias. :)

File: src/Microsoft.OData.Core/UriParser/SemanticAst/ExpandedReferenceSelectItem.cs [hunk 2]
```diff
@@ -9,6 +9,37 @@ namespace Microsoft.OData.UriParser
...
+        /// </summary>
+        /// <param name="handler">An implementation of the handler interface.</param>
+        /// <exception cref="System.ArgumentNullException">Throws if the input handler is null.</exception>
+        public override void HandleWith(SelectItemHandler handler)
+        {
+            handler.Handle(this);
+        }
+
+        public ExpandedCountSelectItem(ODataExpandPath pathToNavigationProperty, IEdmNavigationSource navigationSource) : base(pathToNavigationProperty, navigationSource) { }
+        public ExpandedCountSelectItem(ODataExpandPath pathToNavigationProperty, IEdmNavigationSource navigationSource, FilterClause filterOption, OrderByClause orderByOption, long? topOption, long? skipOption, bool? countOption, SearchClause searchOption) : base(pathToNavigationProperty, navigationSource, filterOption, orderByOption, topOption, skipOption, countOption, searchOption) { }
```
Comment (xuzhg) [inline src/Microsoft.OData.Core/UriParser/SemanticAst/ExpandedReferenceSelectItem.cs]: @joshcomley The spec says " *The $count segment can be appended to a navigation property name or type-cast segment following a navigation property name to return just the count of the related entities. The $filter and $search system query options can be used to limit the number or related entities included in the count.* ",  

So, it seems it only accepts "filterOption" and "SearchOption" in "ExpandedCountSelectItem" ?

Comment (joshcomley) [reply src/Microsoft.OData.Core/UriParser/SemanticAst/ExpandedReferenceSelectItem.cs]: @xuzhg It should only accept filter and search, yes

**Comment by joshcomley** [comment]
I've just realised this issue is still outstanding (have been using my own patched version of OData for a while). What do I need to do to get this merged?

**Comment by xuzhg** [comment]
@joshcomley Thanks for tracking on this. I have fixed some failure test cases and recreate a PR at: #1336. 

We do like this feature to be implemented. However, for your changes, i'd like to change some design and implementation.  Owing to other high priority task, this change is postpone. It seems you are interested it again, maybe we can find some time to sync it.

**Comment by joshcomley** [comment]
Hi @xuzhg, thanks for getting back to me. As it's been a year since these changes, some things have changed, but I have re-applied my changes to the latest version of the code-base. Should I delete this PR and create a new one with the new amendments?

**Comment by joshcomley** [comment]
This PR has today been rebased on master, the changes added in https://github.com/OData/odata.net/pull/1336 have been merged in and I have added support for `$select=Collection/$count` (it already supported `$expand=Collection/$count`

**Comment by KenitoInc** [comment]
@joshcomley 
We merged these 2 PRs #2051 and #2069 
They will be out in v7.9 release. 
Check if they resolve your issue

**Pull request closed without merging**